oxidation/libparsec/crates/crypto/src/rustcrypto/secret.rs

```diff
@@ -101,9 +101,19 @@ impl SecretKey {
     pub fn from_password(password: &str, salt: &[u8]) -> Self {
         let mut key = [0; KEY_SIZE];
 
-        ARGON2
-            .hash_password_into(password.as_bytes(), salt, &mut key)
-            .expect("Invalid salt");
+        if cfg!(feature = "test-unsecure-but-fast-secretkey-from-password") {
+            let password_end = KEY_SIZE.min(password.len());
+
+            key[..password_end].copy_from_slice(&password.as_bytes()[..password_end]);
+
+            let salt_end = (KEY_SIZE - password_end).min(salt.len());
+
+            key[password_end..password_end + salt_end].copy_from_slice(&salt[..salt_end]);
```

Comment:
`password_end + salt_end` seems equal to `KEY_SIZE`:
if `password.len() >= KEY_SIZE`, then `password_end == KEY_SIZE`, so `salt_end == 0` so the sum is equal to `KEY_SIZE`
if `password.len() < KEY_SIZE`, then `salt_end == KEY_SIZE - password_end`, so the sum is also equal to `KEY_SIZE`

Reply:
> `password_end + salt_end` seems equal to `KEY_SIZE`: 
> if `password.len() >= KEY_SIZE`, then `password_end == `KEY_SIZE`, 

That the role of `password_end = min(KEY_SIZE, password.len())`

> so `salt_end == 0` so the sum is equal to `KEY_SIZE` 

Also the case with `salt_end = min(KEY_SIZE - password_end, salt.len())`

> if `password.len() < KEY_SIZE`, then `salt_end == KEY_SIZE - password_end`, so the sum is also equal to `KEY_SIZE`

-----

I'm don't understand this discussion, what do you want from it ?

1. Is it only to deconstruct the logic behind ?
2. Is it to propose / request a change ?

Reply:
My bad, forget about salt.len(), you are right